Kernel/Arch/Processor.h

```diff
@@ -8,18 +8,201 @@
+class ProcessorBase {
+    friend class Processor;
+
+public:
+    template<typename T>
+    T* get_specific()
+    {
+        return static_cast<T*>(m_processor_specific_data[static_cast<size_t>(T::processor_specific_data_id())]);
+    }
+
+    void set_specific(ProcessorSpecificDataID specific_id, void* ptr)
+    {
+        m_processor_specific_data[static_cast<size_t>(specific_id)] = ptr;
+    }
+
+    static bool is_smp_enabled();
+    static void smp_enable();
+    static u32 smp_wake_n_idle_processors(u32 wake_count);
+
+    static void flush_tlb_local(VirtualAddress vaddr, size_t page_count);
+    static void flush_tlb(Memory::PageDirectory const*, VirtualAddress, size_t);
+
+    void early_initialize(u32 cpu);
+    void initialize(u32 cpu);
+    ALWAYS_INLINE static bool is_initialized();
+
+    [[noreturn]] static void halt();
+    void wait_for_interrupt() const;
+    ALWAYS_INLINE static void pause();
+    ALWAYS_INLINE static void wait_check();
+
+    ALWAYS_INLINE static ProcessorT& current();
+    static Processor& by_id(u32);
+
+    ALWAYS_INLINE u32 id() const
+    {
+        // NOTE: This variant should only be used when iterating over all
+        // Processor instances, or when it's guaranteed that the thread
+        // cannot move to another processor in between calling Processor::current
+        // and Processor::id, or if this fact is not important.
+        // All other cases should use Processor::current_id instead!
+        return m_cpu;
+    }
+
+    ALWAYS_INLINE static u32 current_id();
+    ALWAYS_INLINE static bool is_bootstrap_processor();
+    ALWAYS_INLINE bool has_nx() const;
+    ALWAYS_INLINE bool has_pat() const;
+    ALWAYS_INLINE bool has_feature(CPUFeature::Type const& feature) const
+    {
+        return m_features.has_flag(feature);
+    }
+    static StringView platform_string();
+
+    static u32 count()
+    {
+        // NOTE: because this value never changes once all APs are booted,
+        // we can safely bypass loading it atomically.
+        // NOTE: This does not work on aarch64, since the variable is never written.
+        return *g_total_processors.ptr();
+    }
+
+    void enter_trap(TrapFrame& trap, bool raise_irq);
+    void exit_trap(TrapFrame& trap);
+
+    static void flush_entire_tlb_local();
+
+    ALWAYS_INLINE static Thread* current_thread();
+    ALWAYS_INLINE static void set_current_thread(Thread& current_thread);
+    ALWAYS_INLINE static Thread* idle_thread();
+
+    ALWAYS_INLINE static u32 in_critical();
+    ALWAYS_INLINE static void enter_critical();
+    static void leave_critical();
+    void do_leave_critical();
+    static u32 clear_critical();
+    ALWAYS_INLINE static void restore_critical(u32 prev_critical);
+    ALWAYS_INLINE static void verify_no_spinlocks_held()
+    {
+        VERIFY(!ProcessorBase::in_critical());
+    }
+
+    static InterruptsState processor_interrupts_state();
+    static void restore_processor_interrupts_state(InterruptsState);
```

Comment:
The `processor` prefix can now be removed, since this function lives inside the `ProcessorBase` now.